Use all text scorers in the final result to increase scoring signal

escherize opened the pull request:
Fixes #19366

- use the results of all text scorers instead of just the max.
- add tests

## Use all the text-* matcher results

Instead of the max one, there is some signal to be gained by taking all of them, and combining every sort of score (i.e. a map with `:name`, `:score`, `:weight`) right at the end. That helps figure out "Why is it sorted this way", which is a question I was trying to figure out at first.

## Introduces the `prefix-scorer`.

Even though it seems naive, this scorer helps gain more signal, and it is also really fast. In my testing it has helped to boost exact matches too.

escherize committed:
Uses all text scorers in the final result

- instead of just the maximum one
- add tests

escherize committed:
add prefix scorer test + fix text-score-with

escherize committed:
linter fixes

escherize committed:
pass in number of results to find

escherize committed:
refactor test function

escherize committed:
fix linter by removing unused namespace: metabase.util

escherize committed:
limit arity of serialize to 3

escherize committed:
make oss-score and ee-score different things

- They were defined to be exactly the same, but should be different!
- Update some tests that broke when a test function was fixed

escherize committed:
remove extra let

escherize committed:
Merge branch 'master' of github.com:metabase/metabase into add-prefix-scorer

escherize commented on enterprise/backend/test/metabase_enterprise/search/scoring_test.clj:
```diff
@@ -18,67 +21,104 @@
-      (-> (scoring/score-and-result search-string item) :score))))
+  [search-string item]
+  (with-redefs [#_{:clj-kondo/ignore [:deprecated-var]}
+                premium-features/enable-enhancements? (constantly false)]
```
This typo caused me to make a bunch of changes in these tests! `oss-score` and `ee-score` were behaving the same way.

escherize committed:
move rseq back out of sorted-take

escherize committed:
improve test feedback

escherize committed:
force weight of text based scorers always weigh 10

escherize committed:
handle 0 score/weights when normalizing scores

escherize committed:
add nil check

escherize committed:
fix more subtle test differences

escherize committed:
more test fiddling

- still test that :offset and :limit respect limits

escherize committed:
reuse bit->boolean from api collection

escherize commented on src/metabase/search/scoring.clj:
```diff
@@ -138,23 +136,44 @@
+   {:scorer total-occurrences-scorer :name "total-occurrences" :weight 2}
+   {:scorer fullness-scorer :name "fullness" :weight 1}
+   {:scorer prefix-scorer :name "prefix" :weight 1}])
 
```
Sorted these by their weights. Also added prefix-scorer.

tsmacdonald replied:
Adding `name` *almost* makes me want a macro

```
(scorer consecutivity 2) ;; => {:scorer consecutivity-scorer :name "consecutivity" :weight 2}
```

but probably it's cleaner to just accept the repetition and not worry about it

escherize commented on src/metabase/search/scoring.clj:
```diff
@@ -271,31 +275,51 @@
-       {:score 0}))))
+  "Returns a map with the normalized, combined score from relevant-scores as `:score` and `:result`."
+  [raw-search-string result]
+  (let [text-matches     (force-weight (text-scores-with-match raw-search-string result) 10)
```
To disrupt less testing, we keep the text-matches at weight 10. We just return all of them, so we can get a clearer picture of the "why" behind result sorting.

escherize commented on src/metabase/search/scoring.clj:
```diff
@@ -271,31 +275,51 @@
+       :result (serialize result all-scores relevant-scores)}
+      {:score 0})))
+
+(defn compare-score
```
`compare-score` was moved down since it wasn't needed until top-results

escherize committed:
clean up some tests

- filter -> remove
- replace some magic numbers
- revert to testing entire maps instead of names of sorted items

escherize committed:
add test, docstring, and weight

escherize committed:
Merge branch 'master' into add-prefix-scorer

escherize committed:
Merge branch 'master' into add-prefix-scorer

escherize committed:
sort ns requires

escherize committed:
Merge branch 'add-prefix-scorer' of github.com:metabase/metabase into add-prefix-scorer

tsmacdonald commented on src/metabase/search/scoring.clj:
```diff
@@ -81,31 +81,29 @@
+  ;; TODO is pmap over search-result worth it?
+  (let [scores       (for [column      (search-config/searchable-columns-for-model (:model search-result))
+                           {:keys [scorer name weight]
+                            :as   _ws} weighted-scorers
```
Nit, why bother with the `:as _ws` ?

escherize replied:
That is a trick I use to keep the alignment of the r.h.s. of the for comprehension from being too far over to the right

<img width="1060" alt="Screen Shot 2022-10-27 at 4 24 43 PM" src="https://user-images.githubusercontent.com/343288/198409577-508b4c0d-45b9-4200-b5a4-146db4a668e0.png">

If you think it's too weird I'm open to nixing it.

tsmacdonald commented on src/metabase/search/scoring.clj:
```diff
@@ -81,31 +81,29 @@
 
-(defn- text-score-with
+(defn- text-scores-with
   "Scores a search result. Returns a map with the score and other info about the text match,
    if there is one. If there is no match, the score is 0."
```
```suggestion
  "Scores a search result. Returns a vector of score maps, each containing the `:score`
  and other details about the text match, if there is one. If there is no match, a vector with a single map of score 0 will be returned."
```

(or something. And the linebreaking/whitespace will need fixing)

tsmacdonald commented on src/metabase/search/scoring.clj:
```diff
@@ -164,14 +183,13 @@
+    (text-scores-with match-based-scorers
+                      (tokenize (normalize raw-search-string))
+                      result)
+    [{:score 0 :weight 1 :match ""}]))
```
I'm wondering if `text-scores-with` should handle the nil `raw-search-string` case so that its `{:score 0 ...}` result can be returned if there's no search string or if there's no match, instead of having the separate `{:score 0 ...}` here.

I also don't understand why the one in `text-scores-with` has weight 0 and this one has weight 1.

(I have a sinking sensation that if you `git blame` long enough you'll discover I'm the one who wrote it that way in the first place. If so, sorry!)

escherize replied:
Good catch, I'll try having `text-scores-with-match` handle the empty `raw-search-string`. I also have the philosophy that the more stuff someone does the more mistakes they'll make so mistakes aren't such a bad thing! Mine or yours!

tsmacdonald reviewed (approved):
:boat:

metamben commented on src/metabase/search/scoring.clj:
```diff
@@ -271,31 +275,53 @@
+  [scores total]
+  (let [total-found (reduce + (map :weight scores))]
+    (mapv #(update % :weight (fn [weight]
+                               (if (contains? #{nil 0} total-found)
+                                 0
+                                 (* total (/ weight total-found))))) scores)))
```
Wouldn't it be clearer if we created a function before mapping? In the `(contains? #{nil 0} total-found)` case it'd be `(constantly 0)`.

metamben commented on test/metabase/api/search_test.clj:
```diff
@@ -134,7 +134,7 @@
 (def ^:private remove-databases
   "Remove DBs from the results, which is useful since test databases unrelated to this suite can pollute the results"
-  (partial filter #(not= (:model %) "database")))
+  (partial remove #(= (:model %) "database")))
```
👍

metamben commented on test/metabase/api/search_test.clj:
```diff
@@ -215,16 +215,16 @@
-      (is (= 2 (count (search-request-data :crowberto :q "test" :limit "2" :offset "0"))))))
+      (is (>= 2 (count (search-request-data :crowberto :q "test" :limit "2" :offset "0"))))))
   (testing "It offsets matches properly"
     (with-search-items-in-root-collection "test"
       (is (<= 4 (count (search-request-data :crowberto :q "test" :limit "100" :offset "2"))))))
   (testing "It offsets without limit properly"
     (with-search-items-in-root-collection "test"
-      (is (= 5 (count (search-request-data :crowberto :q "test" :offset "2"))))))
+      (is (<= 5 (count (search-request-data :crowberto :q "test" :offset "2"))))))
   (testing "It limits without offset properly"
     (with-search-items-in-root-collection "test"
-      (is (= 2 (count (search-request-data :crowberto :q "test" :limit "2"))))))
+      (is (>= 2 (count (search-request-data :crowberto :q "test" :limit "2"))))))
```
It's not clear to me why we sometimes relax the lower bound, sometimes the upper bound.

escherize replied:
It's related to the change in search ordering.

There an updated search ordering, and `search-request-data` removes databases (for a good reason -- e.g. if you have test databases they will show up in the search results, so the tests would be unstable based on the testing env).

So that is why the results here changed -- I think I can use `=` instead tho, after all.

metamben commented on test/metabase/search/scoring_test.clj:
```diff
@@ -302,3 +317,20 @@
+  (is (= [{:weight 10}]
+         (scoring/force-weight [{:weight 1}] 10)))
+
+  (is (= [{:weight 5N} {:weight 5N}]
```
Wouldn't `5` work instead of `5N`?

escherize committed:
responding to most of the review comments

escherize commented on src/metabase/api/search.clj:
```diff
@@ -440,36 +441,32 @@
        :order-by order-clause})))
 
-(s/defn ^:private search
+(defn ^:private search
```
No, I take those off because my debugger doesn't work with the schema/defn. I'll add it back in. 👍

escherize commented on src/metabase/search/scoring.clj:
```diff
@@ -138,23 +136,44 @@
-    :weight 1/2}
-   {:scorer exact-match-scorer
-    :weight 2}])
+  [{:scorer exact-match-scorer :name "exact-match" :weight 4}
```
Yeah, And I think it does do that. Why are you confused here? Maybe I am missing something

metamben replied:
Just because the original issue seems to about an exact match not getting top score.

escherize commented on src/metabase/search/scoring.clj:
```diff
@@ -244,21 +258,11 @@
-    :name   "model"}])
+  [{:weight 2 :score (pinned-score result) :name "pinned"}
+   {:weight 2 :score (bookmarked-score result) :name "bookmarked"}
+   {:weight 3/2 :score (recency-score result) :name "recency"}
```
I moved the other weights to integers, I think we agree that it's aesthetically pleasing. The reason I could do that is because I force the "text-*" scorers total weight to be exactly 10. Meaning the weights I changed are only relative to each other. But doubling these weights would reduce the signal coming from the "text-*" scorers, so I left them unchanged.

escherize commented on src/metabase/search/scoring.clj:
```diff
@@ -271,31 +275,53 @@
    (weights-and-scores result))
 
+(defn- compute-normalized-score [scores]
+  (let [weight-sum (reduce + (map #(or (:weight %) 0) scores))
```
I noticed that I was doing this weight summing in a few places, and pulled it out into a fn.

```
(defn- sum-weights [weights]
  (reduce
   (fn [acc {:keys [weight] :or {weight 0}}]
     (+ acc weight))
   0
   weights))
```

`:weight` might not be there, but the weights are assumed to be maps.

escherize commented on src/metabase/search/scoring.clj:
```diff
@@ -271,31 +275,53 @@
+(defn force-weight
+  "Reweight `scores` such that the sum of their weights equals `total`, and their proportions do not change."
+  [scores total]
+  (let [total-found (reduce + (map :weight scores))]
```
This is the second place I'm computing the total weight, I've pulled it into a fxn.

escherize committed:
start our zero-score sum check with 0

escherize committed:
do not tokenize / normalize nil raw-search-string

escherize committed:
force equality in basic search test

escherize committed:
Merge remote-tracking branch 'origin/master' into add-prefix-scorer

escherize committed:
modify test to work in dev and test environments

escherize committed:
Merge branch 'master' of github.com:metabase/metabase into add-prefix-scorer

escherize committed:
use display_name in results when appropriate

- This was looking for the _first_ column that had a non-zero score, but
actually we need to consider all relevant columns.
- Uses them to figure out if there is a display name, and if there is,
to use it.
- Coppied over the logic about showing :context from the prior approach

escherize committed:
Merge branch 'master' into add-prefix-scorer

flamber commented:
@escherize Is this going into 45 or 46? If 45, then you need to add `backport` label to this PR.
